SignalR Java SDK throwing UndeliverableException

<!--

More information on our issue management policies can be found here: https://aka.ms/aspnet/issue-policies

Please keep in mind that the GitHub issue tracker is not intended as a general support forum, but for reporting **non-security** bugs and feature requests.

If you believe you have an issue that affects the SECURITY of the platform, please do NOT create an issue and instead email your issue details to secure@microsoft.com. Your report may be eligible for our [bug bounty](https://www.microsoft.com/en-us/msrc/bounty-dot-net-core) but ONLY if it is reported through email.
For other types of questions, consider using [StackOverflow](https://stackoverflow.com).

-->

### Describe the bug
We are encountering a set of crashes from within the SignalR Java SDK. There are two separate issues but both are uncaught exceptions that currently crash our app. The two issues are:
1. com.microsoft.signalr.HubConnection$ConnectionState.lambda$timeoutHandshakeResponse$0
java.util.concurrent.TimeoutException: Timed out waiting for the server to respond to the handshake message.
2. com.microsoft.signalr.HandshakeProtocol.parseHandshakeResponse
HandshakeProtocol.java, line 21
java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $

Note that the issue is not that the exception is being thrown in the first place, but that RxJava delivers it as an UndeliverableException and not through the subject as intended:

https://github.com/dotnet/aspnetcore/blob/bc1ff6a45949c3debc483aae93abd0dfdf97bec1/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java#L1433-L1440

I am wondering if there could be a race between the timeout and `handleHandshake` that could lead to the Subject being completed multiple times since there is no synchronization in place.

I'm also curious why the timeout is handled by creating a new executor instead of just using RxJava's `timeout` operator?

For the time being we are registering an RxJava error handler to log and swallow these exceptions but it would be good if this could be fixed and the errors are being propagated to the Completable that is returned by `HubConnection.start()`

### To Reproduce
<!--
We ❤ code! Point us to a minimalistic repro project hosted in a GitHub repo.
For a repro project, create a new ASP.NET Core project using the template of your your choice, apply the minimum required code to result in the issue you're observing.

We will close this issue if:
- the repro project you share with us is complex. We can't investigate custom projects, so don't point us to such, please.
- if we will not be able to repro the behavior you're reporting
-->
Wasn't able to reproduce this.

### Exceptions (if any)
<!-- 
Include the exception you get when facing this issue
-->
Stacktrace TimeoutException:
```
io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.util.concurrent.TimeoutException: Timed out waiting for the server to respond to the handshake message.
  at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
  at io.reactivex.subjects.CompletableSubject.onError(CompletableSubject.java:131)
  at com.microsoft.signalr.HubConnection$ConnectionState.lambda$timeoutHandshakeResponse$0(HubConnection.java:1428)
  at com.microsoft.signalr.HubConnection$ConnectionState.lambda$timeoutHandshakeResponse$0$HubConnection$ConnectionState(HubConnection.java)
  at com.microsoft.signalr.-$$Lambda$HubConnection$ConnectionState$_fs5XriBgCmY7mCMawuT2CL0Nx4.run(-.java)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:459)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  at java.lang.Thread.run(Thread.java:764)
  Caused by: java.util.concurrent.TimeoutException: Timed out waiting for the server to respond to the handshake message.
    at com.microsoft.signalr.HubConnection$ConnectionState.lambda$timeoutHandshakeResponse$0(HubConnection.java:1428)
    at com.microsoft.signalr.HubConnection$ConnectionState.lambda$timeoutHandshakeResponse$0$HubConnection$ConnectionState(HubConnection.java)
    at com.microsoft.signalr.-$$Lambda$HubConnection$ConnectionState$_fs5XriBgCmY7mCMawuT2CL0Nx4.run(-.java)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:459)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
    at java.lang.Thread.run(Thread.java:764)
```

Stacktrace TimeoutException:
```
io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.lang.RuntimeException: An invalid handshake response was received from the server.
  at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
  at io.reactivex.subjects.CompletableSubject.onError(CompletableSubject.java:131)
  at com.microsoft.signalr.HubConnection$ConnectionState.handleHandshake(HubConnection.java:1407)
  at com.microsoft.signalr.HubConnection.ReceiveLoop(HubConnection.java:437)
  at com.microsoft.signalr.HubConnection.lambda$new$0(HubConnection.java:162)
  at com.microsoft.signalr.HubConnection.lambda$new$0$HubConnection(HubConnection.java)
  at com.microsoft.signalr.-$$Lambda$HubConnection$QTbpx7gje_2sI4dX1CsfAa9dG7g.invoke(-.java:2)
  at com.microsoft.signalr.WebSocketTransport.onReceive(WebSocketTransport.java:75)
  at com.microsoft.signalr.WebSocketTransport.lambda$start$0(WebSocketTransport.java:52)
  at com.microsoft.signalr.WebSocketTransport.lambda$start$0$WebSocketTransport(WebSocketTransport.java)
  at com.microsoft.signalr.-$$Lambda$WebSocketTransport$IuEB82SnLNJc1Av0IiMi3-Bx-v8.invoke(-.java)
  at com.microsoft.signalr.OkHttpWebSocketWrapper$SignalRWebSocketListener.onMessage(OkHttpWebSocketWrapper.java:100)
  at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:338)
  at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:247)
  at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:106)
  at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:293)
  at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.java:195)
  at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.java:519)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  at java.lang.Thread.run(Thread.java:919)
  Caused by: java.lang.RuntimeException: An invalid handshake response was received from the server.
    at com.microsoft.signalr.HubConnection$ConnectionState.handleHandshake(HubConnection.java:1406)
    at com.microsoft.signalr.HubConnection.ReceiveLoop(HubConnection.java:437)
    at com.microsoft.signalr.HubConnection.lambda$new$0(HubConnection.java:162)
    at com.microsoft.signalr.HubConnection.lambda$new$0$HubConnection(HubConnection.java)
    at com.microsoft.signalr.-$$Lambda$HubConnection$QTbpx7gje_2sI4dX1CsfAa9dG7g.invoke(-.java:2)
    at com.microsoft.signalr.WebSocketTransport.onReceive(WebSocketTransport.java:75)
    at com.microsoft.signalr.WebSocketTransport.lambda$start$0(WebSocketTransport.java:52)
    at com.microsoft.signalr.WebSocketTransport.lambda$start$0$WebSocketTransport(WebSocketTransport.java)
    at com.microsoft.signalr.-$$Lambda$WebSocketTransport$IuEB82SnLNJc1Av0IiMi3-Bx-v8.invoke(-.java)
    at com.microsoft.signalr.OkHttpWebSocketWrapper$SignalRWebSocketListener.onMessage(OkHttpWebSocketWrapper.java:100)
    at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:338)
    at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:247)
    at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:106)
    at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:293)
    at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.java:195)
    at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.java:519)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
    at java.lang.Thread.run(Thread.java:919)
    Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:226)
      at com.google.gson.Gson.fromJson(Gson.java:927)
      at com.google.gson.Gson.fromJson(Gson.java:892)
      at com.google.gson.Gson.fromJson(Gson.java:841)
      at com.google.gson.Gson.fromJson(Gson.java:813)
      at com.microsoft.signalr.HandshakeProtocol.parseHandshakeResponse(HandshakeProtocol.java:21)
      at com.microsoft.signalr.HubConnection$ConnectionState.handleHandshake(HubConnection.java:1404)
      at com.microsoft.signalr.HubConnection.ReceiveLoop(HubConnection.java:437)
      at com.microsoft.signalr.HubConnection.lambda$new$0(HubConnection.java:162)
      at com.microsoft.signalr.HubConnection.lambda$new$0$HubConnection(HubConnection.java)
      at com.microsoft.signalr.-$$Lambda$HubConnection$QTbpx7gje_2sI4dX1CsfAa9dG7g.invoke(-.java:2)
      at com.microsoft.signalr.WebSocketTransport.onReceive(WebSocketTransport.java:75)
      at com.microsoft.signalr.WebSocketTransport.lambda$start$0(WebSocketTransport.java:52)
      at com.microsoft.signalr.WebSocketTransport.lambda$start$0$WebSocketTransport(WebSocketTransport.java)
      at com.microsoft.signalr.-$$Lambda$WebSocketTransport$IuEB82SnLNJc1Av0IiMi3-Bx-v8.invoke(-.java)
      at com.microsoft.signalr.OkHttpWebSocketWrapper$SignalRWebSocketListener.onMessage(OkHttpWebSocketWrapper.java:100)
      at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:338)
      at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:247)
      at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:106)
      at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:293)
      at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.java:195)
      at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.java:519)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
      at java.lang.Thread.run(Thread.java:919)
```

### Further technical details
- SignalR Java SDK 5.0.3

--- a/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java
+++ b/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java
@@ -349,7 +349,10 @@ public class HubConnection implements AutoCloseable {
                     this.state.unlock();
                 }
 
-                localStart.onError(error);
+                // this should be thread-safe as there shouldn't be any active calls to the accessTokenProvider
+                if (!(localStart.hasComplete() || localStart.hasThrowable())) {
+                    localStart.onError(error);
+                }
             });
         } finally {
             this.state.lock.unlock();
@@ -1413,14 +1416,14 @@ public class HubConnection implements AutoCloseable {
                     handshakeResponse = HandshakeProtocol.parseHandshakeResponse(handshakeResponseString);
                 } catch (RuntimeException ex) {
                     RuntimeException exception = new RuntimeException("An invalid handshake response was received from the server.", ex);
-                    handshakeResponseSubject.onError(exception);
+                    errorHandshake(exception);
                     throw exception;
                 }
                 if (handshakeResponse.getHandshakeError() != null) {
                     String errorMessage = "Error in handshake " + handshakeResponse.getHandshakeError();
                     logger.error(errorMessage);
                     RuntimeException exception = new RuntimeException(errorMessage);
-                    handshakeResponseSubject.onError(exception);
+                    errorHandshake(exception);
                     throw exception;
                 }
                 handshakeReceived = true;
@@ -1431,12 +1434,7 @@ public class HubConnection implements AutoCloseable {
         public void timeoutHandshakeResponse(long timeout, TimeUnit unit) {
             handshakeTimeout = Executors.newSingleThreadScheduledExecutor();
             handshakeTimeout.schedule(() -> {
-                // If onError is called on a completed subject the global error handler is called
-                if (!(handshakeResponseSubject.hasComplete() || handshakeResponseSubject.hasThrowable()))
-                {
-                    handshakeResponseSubject.onError(
-                        new TimeoutException("Timed out waiting for the server to respond to the handshake message."));
-                }
+                errorHandshake(new TimeoutException("Timed out waiting for the server to respond to the handshake message."));
             }, timeout, unit);
         }
 
@@ -1475,6 +1473,18 @@ public class HubConnection implements AutoCloseable {
             }
 
             return handlers.get(0).getTypes();
+        }
+
+        private void errorHandshake(Exception error) {
+            lock.lock();
+            try {
+                // If onError is called on a completed subject the global error handler is called
+                if (!(handshakeResponseSubject.hasComplete() || handshakeResponseSubject.hasThrowable())) {
+                    handshakeResponseSubject.onError(error);
+                }
+            } finally {
+                lock.unlock();
+            }
         }
     }
 

--- a/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java
+++ b/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java
@@ -349,10 +349,7 @@ public class HubConnection implements AutoCloseable {
                     this.state.unlock();
                 }
 
-                // this should be thread-safe as there shouldn't be any active calls to the accessTokenProvider
-                if (!(localStart.hasComplete() || localStart.hasThrowable())) {
-                    localStart.onError(error);
-                }
+                localStart.onError(error);
             });
         } finally {
             this.state.lock.unlock();

--- a/src/SignalR/clients/java/signalr/test/src/main/java/com/microsoft/signalr/HubConnectionTest.java
+++ b/src/SignalR/clients/java/signalr/test/src/main/java/com/microsoft/signalr/HubConnectionTest.java
@@ -17,6 +17,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
 
 import ch.qos.logback.classic.spi.ILoggingEvent;
 import io.reactivex.rxjava3.core.Completable;
@@ -29,6 +30,7 @@ import io.reactivex.rxjava3.subjects.PublishSubject;
 import io.reactivex.rxjava3.subjects.ReplaySubject;
 import io.reactivex.rxjava3.subjects.SingleSubject;
 
+@ExtendWith({RxJavaUnhandledExceptionsExtensions.class})
 class HubConnectionTest {
     private static final String RECORD_SEPARATOR = "\u001e";
     private static final Type booleanType = (new TypeReference<Boolean>() { }).getType();

--- a/src/SignalR/clients/java/signalr/test/src/main/java/com/microsoft/signalr/RxJavaUnhandledExceptionsExtensions.java
+++ b/src/SignalR/clients/java/signalr/test/src/main/java/com/microsoft/signalr/RxJavaUnhandledExceptionsExtensions.java
@@ -1,0 +1,41 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+package com.microsoft.signalr;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+import io.reactivex.rxjava3.plugins.RxJavaPlugins;
+
+// Use by adding "@ExtendWith({RxJavaUnhandledExceptionsExtensions.class})" to a test class
+class RxJavaUnhandledExceptionsExtensions implements BeforeAllCallback, AfterAllCallback {
+    private final BlockingQueue<Throwable> errors = new LinkedBlockingQueue<Throwable>();
+
+    @Override
+    public void beforeAll(final ExtensionContext context) {
+        RxJavaPlugins.setErrorHandler(error -> {
+            errors.put(error);
+        });
+    }
+
+    @Override
+    public void afterAll(final ExtensionContext context) {
+        if (errors.size() != 0) {
+            String RxErrors = "";
+            for (final Throwable throwable : errors) {
+                StringWriter stringWriter = new StringWriter();
+                PrintWriter printWriter = new PrintWriter(stringWriter);
+                throwable.printStackTrace(printWriter);
+                RxErrors += String.format("%s\n", stringWriter.toString());
+            }
+            throw new RuntimeException(RxErrors);
+        }
+    }
+}

